projects/vdk-plugins/vdk-ipython/tests/test_plugin.py

```diff
@@ -169,3 +180,66 @@ def test_extension_with_pure_sql_job(ip, tmpdir):
     assert "2021-01-01  GOOG          123" in (
         ip.get_ipython().getoutput("! " "vdk " "sqlite-query -q 'SELECT * FROM stocks'")
     )
+
+
+def test_finalise(ip):
+    ip.run_line_magic(magic_name="reload_VDK", line="")
+
+    ip.get_ipython().run_cell("job_input = VDK.get_initialized_job_input()")
+
+    # check whether finalize_job is in the output before calling finalize()
+    assert (
+        "finalize_job"
+        not in ip.get_ipython()
+        .run_cell("VDK.job._plugin_registry.hook().__dict__")
+        .result
+    )
+
+    ip.get_ipython().run_cell("VDK.finalize()")
+
+    # check whether finalize_job is in the output after calling finalize()
+    assert (
+        "finalize_job"
+        in ip.get_ipython().run_cell("VDK.job._plugin_registry.hook().__dict__").result
+    )
+
+
+def test_get_initialized_job_input_multiple_times(ip):
+    ip.run_line_magic(magic_name="reload_VDK", line="")
+
+    ip.get_ipython().run_cell("job_input = VDK.get_initialized_job_input()")
+    first_result = ip.get_ipython().run_cell("job_input").result
+    ip.get_ipython().run_cell("job_input = VDK.get_initialized_job_input()")
+
+    assert first_result == ip.get_ipython().run_cell("job_input").result
+
+
+def test_get_initialized_job_input_after_finalize(ip):
+    ip.run_line_magic(magic_name="reload_VDK", line="")
+
+    ip.get_ipython().run_cell("job_input = VDK.get_initialized_job_input()")
+    ip.get_ipython().run_cell("VDK.finalize()")
+    ip.get_ipython().run_cell("job_input = VDK.get_initialized_job_input()")
+    ip.get_ipython().run_cell("VDK.finalize()")
+    ip.get_ipython().run_cell("job_input = VDK.get_initialized_job_input()")
+
+    assert ip.user_global_ns["job_input"] is not None
```

Comment:
the test was supposed to verify that no errors occur if we want to get job_input for the second time. It does not test whether it is different or not. I might add a test for checking the difference , too.

Reply:
There's no need to add multiple tests. We can use multiple asserts verifications for different things in the same test if the setup(arrange) and the code under test (act) are the same.

Otherwise we have a lot of code duplication in the test which is not good.

Reply:
alternatively you can extract the common code in a method and call it from both tests, but if the assert part is not complex simple overkill